src/shared/sam/cli/samCliInfo.ts

```diff
@@ -15,7 +16,22 @@ export interface SamCliInfoResponse {
 }
 
 export class SamCliInfoInvocation {
-    public constructor(private readonly invoker: SamCliProcessInvoker = new DefaultSamCliProcessInvoker()) {}
+    private readonly invoker: SamCliProcessInvoker
+    public constructor(params: {
+        invoker?: SamCliProcessInvoker
+        preloadedConfig?: SamCliConfiguration
+        locationProvider?: { getLocation(): Promise<string | undefined> }
+    }) {
+        if (params.invoker) {
+            this.invoker = params.invoker
+        } else if (params.preloadedConfig) {
+            this.invoker = new DefaultSamCliProcessInvoker({ preloadedConfig: params.preloadedConfig })
+        } else if (params.locationProvider) {
+            this.invoker = new DefaultSamCliProcessInvoker({ locationProvider: params.locationProvider })
+        } else {
+            throw new Error('Invalid constructor args for DefaultSamCliProcessInvoker')
```

Comment:
Are `params.preloadedConfig` and `params.locationProvider` mutually exclusive? If so we should throw if they are both given. If not... yikes :O

Reply:
The main idea is that this gives us a workaround to not break things that are already in place (they can just pass their `preloadedConfig` directly in). I acknowledge that this is extremely lazy; would you rather see safeguards or would you rather me convert to something more uniform?

Reply:
So it's intended that both may be passed, but only the first one is used to initialize `DefaultSamCliProcessInvoker` ?

Reply:
Yes. As it stands:
* the `invoker` is for tests where we need to swap out the invoker for a fake
* the `preloadedConfig` represents the existing method of passing in a `SamCliConfiguration`: https://github.com/aws/aws-toolkit-vscode/pull/1465/files#diff-3536918fd4d056af1d3fbe5546dd966841ad1a8e3332ab832f759923913129d6L34
* the `locationProvider` is primarily used for validation, where we can provide a `locationProvider` on the fly instead of using the default location provider, which would create a circular dependency.